### PR TITLE
Fluid simulation example with double buffering

### PR DESCRIPTION
**Pull request by iwoplaza**

### CODE REVIEW

File: apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+const isValidCoord = wgsl.fn()`(x: i32, y: i32) -> bool {
+  return
+    x < ${gridSizeData} &&
+    x >= 0 &&
+    y < ${gridSizeData} &&
+    y >= 0;
+}`;
+
+const getCell = wgsl.fn()`(x: i32, y: i32) -> vec4f {
+  let index = x + y * ${gridSizeData};
```
Comment (mhawryluk) [inline apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts]: this line appears a lot, so I think it's worth extracting to a constant:
```const index = wgsl`x + y * ${gridSizeData}`;```

File: apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts [hunk 2]
```diff
@@ -0,0 +1,768 @@
...
+  return least_cost_dir;
+}`;
+
+const mainInitWorld = wgsl.fn()`(x: i32, y: i32) {
+  let index = x + y * ${gridSizeData};
+  
+  var value = vec4f(0., 0., 0., 0.);
+
+  let center = vec2f(f32(${gridSizeData}), f32(${gridSizeData})) / 2.;
+  let topCenter = vec2f(f32(${gridSizeData}) / 2., f32(${gridSizeData}));
```
Comment (mhawryluk) [inline apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts]: everything else is snake_case

File: apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts [hunk 3]
```diff
@@ -0,0 +1,768 @@
...
+    least_cost = n_cost;
+  }
+
+  return least_cost_dir;
+}`;
+
+const mainInitWorld = wgsl.fn()`(x: i32, y: i32) {
+  let index = x + y * ${gridSizeData};
+  
+  var value = vec4f(0., 0., 0., 0.);
```
Comment (mhawryluk) [inline apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts]: you can initialize vectors with just one value: `vec4f(0.);`

File: apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts [hunk 4]
```diff
@@ -0,0 +1,768 @@
...
+  { x: 0.5, y: 0, width: 1, height: 0.1, enabled: true }, // floor
+];
+
+function obstaclesToConcrete(): Parsed<BoxObstacle>[] {
+  return obstacles.map(({ x, y, width, height, enabled }) => ({
+    center: [Math.round(x * gridSize), Math.round(y * gridSize)],
+    size: [
+      Math.round(width * gridSize),
+      Math.round(height * gridSize),
+    ] as const,
```
Comment (mhawryluk) [inline apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts]:
```suggestion
    ],
```
 the compiler complains about this `as const` for me in the code editor view

File: apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts [hunk 5]
```diff
@@ -0,0 +1,768 @@
...
+    // left column
+    for (var y = next_min_y; y <= next_max_y; y += 1) {
+      let new_vel = ${computeVelocity}(next_min_x - 1, y);
+      ${setVelocity}(next_min_x - 1, y, new_vel);
+    }
+    
+    // right column
+    for (var y = max(1, next_min_y); y <= min(next_max_y, ${gridSizeData} - 2); y += 1) {
+      let new_vel = ${computeVelocity}(next_max_x + 2, y);
+      // let new_vel = vec2f(0, 1.);
```
Comment (mhawryluk) [inline apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts]: ✂️

File: apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts [hunk 6]
```diff
@@ -0,0 +1,768 @@
...
+ * "Particle System in WebGPU" by Benedikt Peter
+ */
+const rand01 = wgsl.fn('rand01')`() -> f32 {
+  ${randSeed}.x = fract(cos(dot(${randSeed}, vec2<f32>(23.14077926, 232.61690225))) * 136.8168);
+  ${randSeed}.y = fract(cos(dot(${randSeed}, vec2<f32>(54.47856553, 345.84153136))) * 534.7645);
+  return ${randSeed}.y;
+}`;
+
+type GridData = typeof GridData;
+const GridData = arrayOf(vec4f, MAX_GRID_SIZE ** 2);
```
Comment (mhawryluk) [inline apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts:87]: please add a comment describing what x, y, z, w represent

File: apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts [hunk 7]
```diff
@@ -0,0 +1,768 @@
...
+      });
+    },
+  };
+}
+
+const even = makePipelines(gridAlphaBuffer, gridBetaBuffer);
+const odd = makePipelines(gridBetaBuffer, gridAlphaBuffer);
+
+let primary = even;
+const paused = false;
```
Comment (mhawryluk) [inline apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts]: not useful when it's always false ✂️

File: apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts (hunk 3 above)
Comment (reczkok) [reply apps/wigsill-examples/src/examples/simulation/fluid-double-buffering.ts]: I think `vec4f()` works too

**Pull request merged**